designs/0019-variadic_odes.md

```diff
@@ -0,0 +1,173 @@
+  side (`f`). Can be any non-function type.
+
+There is no way to optionally only provide some of the `rel_tol`, `abs_tol`, or
+`max_num_steps` arguments. We can keep the current defaults (`1e-6`, `1e-6`,
+`1e6`) for `ode_bdf`. 
+
+There are a few changes from the previous implementation.
+
+1. `theta`, `x_real`, and `x_int` are no longer broken out. All parameters,
+real data, and integer data should be passed through as a trailing argument
+(with no rules on naming).
+2. The state is written in terms of `vector`s instead of real arrays. ODE states
+are mathematically written and expressed as vectors, so given the interface is
+changing we might as well change this to be consistent.
+3. I shorted the arguments `initial_state` and `initial_time` to `y0` and `t0`.
+These are standard ways to refer to these quantities.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation of these features is relatively straightforward given the
+code from the
+[reduce_sum](https://github.com/stan-dev/design-docs/blob/master/designs/0017-reduce_sum.md)
+design doc and the existing ODE infrastructure. There is not any significant new
+technology needed.
+
+The old `integrate_ode_bdf`, `integrate_ode_adams`, and `integrate_ode_rk45`
+functions should be deprecated with the introduction of these new functions.
+
+They can remain in the code to provide compatibility to models written with the
+old interfaces. These functions can all be written to wrap around the new
+interfaces so there will not be much technical debt accrued maintaining them.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+I guess a drawback is that we should update as much of the existing ODE
+documentation we can to point at the new interfaces. The old interfaces could
+remain for backwards compatibility though, so if we miss anything it is okay.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The most obvious alternatives are:
+
+* Do nothing (technically everything done here could be done in the old
+interfaces).
+
+I believe the current interface is difficult to use and should be replaced
+as soon as we can.
+
+* Change the scope of the functions block to include variables defined in
+the `data`, `transformed data`, `parameters`, and `transformed parameters`
+block.
+
+I'm not opposed to this, but there hasn't been any discussion of it and it is
+probably harder than the proposal here. This also isn't incompatible with the
+current proposal, though it would probably require some extra work.
+
+* Wait until closures are available so that the ODE right hand side can be
+defined by a closure (which would then have the scope to use any variable
+that could otherwise be passed to through as a right hand side argument here).
+
+I'm not opposed to this either, and this also isn't incompatible with the
+current proposal, though it would probably require some extra work.
```

Comment:
Na, we should not wait here. This is great stuff and will make things easier for users. It may actually make things faster as fewer copies will be needed due to avoiding the unpacking step in the RHS. Moreover, we also avoid that *unpacked data* will be cast to a `var` within the ODE RHS, which is probably what happens at the moment for more complicated models.

Reply:
Weeeeee